### PR TITLE
improvement: rewrite ADK adapter to use Portkey Responses API

### PR DESCRIPTION
**Pull request by viraj-s15**
### Description

* **Core Migration:** Migrated `portkey_ai/integrations/adk.py` from Chat Completions to the **Portkey Responses API** (`responses.create`), replacing the entire request/response translation layer.
* **Reasoning/Thinking Support:** Added support for mapping ADK `thinking_budget` to Responses `reasoning.effort` (low/medium/high) with automatic summary and encrypted content passthrough.
* **Streaming Refactor:** Rewrote streaming from accumulation-based chunk processing (~120 lines) to clean **event-driven handling** (~30 lines) using Responses stream events:`response.reasoning_text.delta`, `response.output_text.delta` and `response.completed`

* **Testing & Documentation:** Updated unit tests and added a streaming+thinking example (`examples/adk_streaming_thinking_usage.py`).

### Motivation

This discussion: https://discord.com/channels/1143393887742861333/1462080485583749318/1462080485583749318

TL;DR: The **Portkey Responses API** has native reasoning support, structured streaming events, and a unified interface across acrooss multiple providers.

### Gemini reasoning not surfaced
Gemini reasoning might not surfaced via Portkey Responses endpoint, when using `@vertex-ai/gemini-3-flash-preview` with reasoning enabled, the model consumes reasoning tokens internally (usage.output_tokens_details.reasoning_tokens > 0) but Portkey does not return reasoning output items for Gemini. This seems to be occurring in LiteLLM too ([BerriAI/litellm#10227](https://github.com/BerriAI/litellm/issues/10227)), also found this discussion in the google forums ([discuss.ai.google.dev](https://discuss.ai.google.dev/t/gemini-2-5-pro-using-thought-tokens-but-intermittently-not-returning-thoughts/111894)). The adapter will surface Gemini reasoning automatically once it starts being including these items in the response.

### Anthropic streaming + reasoning thoughts not surfaced
When using `@anthropic/claude-sonnet-4-6` (or other Claude models) with reasoning enabled in streaming mode, `response.reasoning_text.delta` events are not emitted by Portkey's Responses API. Non-streaming reasoning works correctly (both `thought` and `text` items are returned). This might be  limitation in Portkey Anthropic adapter, it translates Responses API requests to Anthropic Chat Completions internally, and the thinking deltas are not being re mapped to Responses SSE events during streaming. The adapter code already handles `response.reasoning_text.delta` events, so Claude streaming thoughts will surface automatically once Portkey's adapter starts emitting them.

### CODE REVIEW

**Comment by narengogi** [comment]
thankyou @viraj-s15 we'll get this tested and reviewed

**Comment by viraj-s15** [comment]
> thankyou @viraj-s15 we'll get this tested and reviewed

Thanks for taking the time to review, will address any review comments

File: portkey_ai/integrations/adk.py [hunk 1]
```diff
@@ -9,220 +9,81 @@
...
-)
-import json
+from typing import TYPE_CHECKING, Any, AsyncGenerator, Iterable, Optional
 import base64
+import json
 import logging
 
 from portkey_ai import AsyncPortkey
 
 logger = logging.getLogger("portkey_ai.integrations.adk")
```
[the remark this answers is not in the document]
Comment (viraj-s15) [reply portkey_ai/integrations/adk.py]: good catch, I added them for debugging then forgot

File: portkey_ai/integrations/adk.py [hunk 2]
```diff
@@ -9,220 +9,81 @@
...
-        self.prompt_tokens = prompt_tokens
-        self.completion_tokens = completion_tokens
-        self.total_tokens = total_tokens
-
-
 def _safe_json_serialize(obj: Any) -> str:
     try:
         return json.dumps(obj, ensure_ascii=False)
-    except (TypeError, OverflowError):
+    except (TypeError, ValueError):
```
[the remark this answers is not in the document]
Comment (viraj-s15) [reply portkey_ai/integrations/adk.py]: Accidental regression, adding this back

File: portkey_ai/integrations/adk.py [hunk 3]
```diff
@@ -246,8 +107,130 @@ def _schema_to_dict(schema: Any) -> dict:
...
+        if not function_response:
+            continue
+        tool_outputs.append(
+            {
+                "type": "function_call_output",
+                "call_id": getattr(function_response, "id", None),
+                "output": _safe_json_serialize(
+                    getattr(function_response, "response", None)
+                ),
+            }
```
[the remark this answers is not in the document]
Comment (viraj-s15) [reply portkey_ai/integrations/adk.py:196]: Added an explicit ValueError if FunctionResponse.id is missing, so the caller gets a clear error message instead of a 400 from the API.

File: portkey_ai/integrations/adk.py [hunk 4]
```diff
@@ -246,8 +107,130 @@ def _schema_to_dict(schema: Any) -> dict:
...
+                {
+                    "type": "function_call",
+                    "id": getattr(function_call, "id", None),
+                    "call_id": getattr(function_call, "id", None)
+                    or getattr(function_call, "name", "function_call"),
+                    "name": getattr(function_call, "name", None),
+                    "arguments": _safe_json_serialize(
+                        getattr(function_call, "args", None)
+                    ),
+                }
```
[the remark this answers is not in the document]
Comment (viraj-s15) [reply portkey_ai/integrations/adk.py:229]: Same approach, added explicit ValueErrors for both FunctionCall.name and FunctionCall.id

File: portkey_ai/integrations/adk.py [hunk 5]
```diff
@@ -426,196 +279,279 @@ def _get_completion_inputs(
...
-        # Extract system instruction role preference before BaseLlm init
+
         sys_role = str(kwargs.pop("system_role", "developer")).lower()
-        self._system_role = (
+
+        base_kwargs = {k: v for k, v in kwargs.items() if k != "model"}
+        try:
+            super().__init__(model=model, **base_kwargs)  # type: ignore[misc,call-arg]  # BaseLlm may be the stub; real BaseLlm signature varies by adk version
+        except TypeError:
+            super().__init__()  # type: ignore[misc,call-arg]  # fallback for older BaseLlm that takes no args
```
[the remark this answers is not in the document]
Comment (viraj-s15) [reply portkey_ai/integrations/adk.py]: Moved the client kwarg extraction (virtual_key, base_url, etc.) before base_kwargs are created so they're no longer accidentally passed to BaseLlm

File: portkey_ai/integrations/adk.py [hunk 6]
```diff
@@ -426,196 +279,279 @@ def _get_completion_inputs(
...
+            "model": self.model,
+            "input": input_items,
+            **({"tools": tools} if tools else {}),
+            **({"text": text_config} if text_config else {}),
+            **({"instructions": instructions} if instructions else {}),
+        }
+        if reasoning:
+            response_args["reasoning"] = reasoning
+            response_args.setdefault("include", ["reasoning.encrypted_content"])
+        response_args.update(self._additional_args)
```
[the remark this answers is not in the document]
Comment (viraj-s15) [reply portkey_ai/integrations/adk.py:611]: Fixed, but went with a simpler approach than what copilot suggested. Instead of popping include from _additional_args before the update and then iterating/deduping both lists, we just do the update() as is and then verify reasoning.encrypted_content is still present in the final include list, re adding it if something goes wrong. The approach copilot suggested is better if adapter ever needs to protect multiple required include entries, dont think its required right now, can add if needed

**Review by narengogi — DISMISSED** [review]

File: portkey_ai/integrations/adk.py [hunk 7]
```diff
@@ -426,196 +293,290 @@ def _get_completion_inputs(
...
+            elif event_type == "response.output_text.delta":
+                delta = getattr(event, "delta", "")
+                if delta:
+                    yield _parts_to_llm_response(
+                        [_build_text_part(delta)],
+                        is_partial=True,
+                    )
+
+            # Function-call deltas are not yielded as partials; the full
+            # function_call item is emitted via the response.completed event.
```
Comment (kiran-4444) [inline portkey_ai/integrations/adk.py]: This assumption isn't correct. In the responses API streaming, function calls (for Gemini models) are delivered via `response.output_item.added`, `response.function_call_arguments.delta`, and not `response.output_item.done`. Right now, function calls are silently lost as we're trying to accept function calls from `done` event. We need to track function calls from the other events as well and merge them into the final response if they're missing from `response.completed`.

**Comment by kiran-4444** [comment]
Great job on the PR @viraj-s15 🙇 I have one concern and I've addressed it in the review comment. Let me know what you think of it. I tried your changes locally, all the changes worked except streaming tool calling. 

Once we get over this, we should be good to merge.

File: portkey_ai/integrations/adk.py (hunk 7 above)
Comment (viraj-s15) [reply portkey_ai/integrations/adk.py]: Hey, good catch, I seemed to have forgotten about tool calls here, I’ve fixed it, we now capture function calls from the streaming events (response.output_item.added, response.function_call_arguments.delta, and response.function_call_arguments.done) and merge them into the final response if they’re missing. Added tests covering some edge cases around this and have also updated the [example script](https://github.com/Portkey-AI/portkey-python-sdk/pull/421/changes#diff-707b452081a5b44329270fe4cc41a124708c540fbd67bcab44c77771235031bd) to also include a tool call so no one else misses this after me.

**Comment by viraj-s15** [comment]
> Great job on the PR @viraj-s15 🙇 I have one concern and I've addressed it in the review comment. Let me know what you think of it. I tried your changes locally, all the changes worked except streaming tool calling.
> 
> Once we get over this, we should be good to merge.

Thanks for the review, your concern is indeed valid, have addressed it, ready for another look when you get the chance 🫡 .

**Comment by kiran-4444** [comment]
Thanks again for working on fixing the streaming function call issue @viraj-s15. I've tested the changes locally and everything is working as expected. This PR can be merged now 🚀

**Review by kiran-4444 — APPROVED** [review]

**Review by narengogi — APPROVED** [review]
